File: services/knowledge-graph/src/knowledge_graph_mcp/tools/knowledge.py
"""get_knowledge and review_knowledge tools."""

import json
from typing import Any, Callable, Awaitable
from knowledge_graph_mcp.db import KnowledgeDB


async def _get_knowledge_impl(
    query: str,
    db: KnowledgeDB,
    embed_fn: Callable[[str, str, str], Awaitable[list[float]]],
    project_id: str,
    location: str,
    category: str | None = None,
    scope: str | None = None,
    tags: list[str] | None = None,
    limit: int = 10,
) -> list[dict[str, Any]]:
    """Semantic search over knowledge base with scope/tag filtering."""
    embedding = await embed_fn(query, project_id, location)

    conditions = ["(scope = $2 OR scope = 'global')"]
    params: list = [str(embedding), scope or "global"]
    param_idx = 3

    if category:
        conditions.append(f"category = ${param_idx}")
        params.append(category)
        param_idx += 1

    if tags:
        conditions.append(f"tags && ${param_idx}::text[]")
        params.append(tags)
        param_idx += 1

    where = " AND ".join(conditions)
    params.append(limit)

    sql = f"""
        SELECT id, category, key, content, scope, tags
        FROM knowledge_base
        WHERE {where}
        ORDER BY embedding <=> $1::vector
        LIMIT ${param_idx}
    """
    return await db.query(sql, params)


async def _review_knowledge_impl(
    candidate_id: int,
    action: str,
    reviewer: str,
    db: KnowledgeDB,
    embed_fn: Callable | None = None,
    project_id: str | None = None,
    location: str | None = None,
) -> str:
    """Approve or reject a knowledge candidate."""
    if action not in ("approve", "reject"):
        return json.dumps({"error": "Action must be 'approve' or 'reject'"})

    if action == "reject":
        await db.execute(
            "UPDATE knowledge_candidates SET status = 'rejected', reviewed_by = $1, reviewed_at = NOW() WHERE id = $2",
            [reviewer, candidate_id],
        )
        return json.dumps({"status": "rejected", "candidate_id": candidate_id})

    rows = await db.query("SELECT * FROM knowledge_candidates WHERE id = $1", [candidate_id])
    if not rows:
        return json.dumps({"error": f"Candidate {candidate_id} not found"})

    candidate = rows[0]
    embedding = await embed_fn(candidate["content"], project_id, location)

    await db.execute(
        """INSERT INTO knowledge_base (category, key, content, embedding, scope, tags, source)
           VALUES ($1, $2, $3, $4::vector, $5, $6::text[], 'approved_candidate')
           ON CONFLICT (key) DO UPDATE SET content = $3, embedding = $4::vector, scope = $5, tags = $6::text[], updated_at = NOW()""",
        [candidate["category"], candidate["key"], candidate["content"],
         str(embedding), candidate["scope"], candidate["tags"]],
    )
    await db.execute(
        "UPDATE knowledge_candidates SET status = 'approved', reviewed_by = $1, reviewed_at = NOW() WHERE id = $2",
        [reviewer, candidate_id],
    )
    return json.dumps({"status": "approved", "candidate_id": candidate_id, "key": candidate["key"]})


def register_knowledge_tools(mcp, get_db, get_settings, embed_fn):
    @mcp.tool()
    async def get_knowledge(
        query: str,
        category: str | None = None,
        scope: str | None = None,
        tags: list[str] | None = None,
        limit: int = 10,
    ) -> str:
        """Semantic search over knowledge base. Auto-filters by scope + global. Optionally filter by category and tags."""
        s = get_settings()
        results = await _get_knowledge_impl(
            query, get_db(), embed_fn, s.gcp_project_id, s.gcp_location,
            category=category, scope=scope, tags=tags, limit=limit,
        )
        return json.dumps(results, default=str)

    @mcp.tool()
    async def review_knowledge(candidate_id: int, action: str, reviewer: str) -> str:
        """Approve or reject a pending knowledge candidate. On approve, entry is embedded and added to knowledge_base."""
        s = get_settings()
        return await _review_knowledge_impl(
            candidate_id, action, reviewer, get_db(), embed_fn, s.gcp_project_id, s.gcp_location,
        )

File: services/knowledge-graph/src/tools/knowledge.ts
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { sql, eq, and, inArray } from "drizzle-orm";
import type { Db } from "../db/index.js";
import type { Config } from "../config.js";
import { embedText } from "../embeddings.js";
import { knowledgeBase, knowledgeCandidates } from "../db/schema.js";

export function registerKnowledgeTools(
  server: McpServer,
  db: Db,
  config: Config,
) {
  server.tool(
    "get_knowledge",
    "Semantic search over the knowledge base with optional filters",
    {
      query: z.string().describe("Text to search for semantically"),
      category: z.string().optional().describe("Filter by category"),
      scope: z.string().optional().describe("Filter by scope"),
      tags: z.array(z.string()).optional().describe("Filter by tags"),
      limit: z.number().optional().default(10).describe("Max results"),
    },
    async ({ query, category, scope, tags, limit }) => {
      const embedding = await embedText(
        query,
        config.gcpProjectId,
        config.gcpLocation,
      );
      const vectorStr = `[${embedding.join(",")}]`;

      const conditions: ReturnType<typeof eq>[] = [];
      if (scope) {
        conditions.push(inArray(knowledgeBase.scope, [scope, "global"]));
      }
      if (category) {
        conditions.push(eq(knowledgeBase.category, category));
      }
      if (tags && tags.length > 0) {
        conditions.push(
          sql`${knowledgeBase.tags} && ARRAY[${sql.join(
            tags.map((t) => sql`${t}`),
            sql`, `,
          )}]::text[]`,
        );
      }

      const results = await db
        .select({
          id: knowledgeBase.id,
          category: knowledgeBase.category,
          key: knowledgeBase.key,
          content: knowledgeBase.content,
          scope: knowledgeBase.scope,
          tags: knowledgeBase.tags,
        })
        .from(knowledgeBase)
        .where(conditions.length > 0 ? and(...conditions) : undefined)
        .orderBy(sql`embedding <=> ${vectorStr}::vector`)
        .limit(limit ?? 10);

      return {
        content: [
          { type: "text" as const, text: JSON.stringify(results, null, 2) },
        ],
      };
    },
  );

  server.tool(
    "review_knowledge",
    "Approve or reject a knowledge candidate",
    {
      candidate_id: z.number().describe("ID of the candidate to review"),
      action: z.enum(["approve", "reject"]).describe("Action to take"),
      reviewer: z.string().describe("Name of the reviewer"),
    },
    async ({ candidate_id, action, reviewer }) => {
      if (action === "reject") {
        await db
          .update(knowledgeCandidates)
          .set({
            status: "rejected",
            reviewedBy: reviewer,
            reviewedAt: new Date(),
          })
          .where(eq(knowledgeCandidates.id, candidate_id));

        return {
          content: [
            {
              type: "text" as const,
              text: JSON.stringify({ status: "rejected", candidate_id }),
            },
          ],
        };
      }

      // Approve: fetch candidate, embed, upsert into knowledge_base
      const [candidate] = await db
        .select()
        .from(knowledgeCandidates)
        .where(eq(knowledgeCandidates.id, candidate_id));

      if (!candidate) {
        return {
          content: [
            {
              type: "text" as const,
              text: JSON.stringify({ error: "Candidate not found" }),
            },
          ],
        };
      }

      const embedding = await embedText(
        candidate.content,
        config.gcpProjectId,
        config.gcpLocation,
      );

      await db
        .insert(knowledgeBase)
        .values({
          category: candidate.category,
          key: candidate.key,
          content: candidate.content,
          embedding,
          scope: candidate.scope,
          tags: candidate.tags,
          source: "learned",
        })
        .onConflictDoUpdate({
          target: knowledgeBase.key,
          set: {
            content: candidate.content,
            embedding,
            category: candidate.category,
            scope: candidate.scope,
            tags: candidate.tags,
            updatedAt: new Date(),
          },
        });

      await db
        .update(knowledgeCandidates)
        .set({
          status: "approved",
          reviewedBy: reviewer,
          reviewedAt: new Date(),
        })
        .where(eq(knowledgeCandidates.id, candidate_id));

      return {
        content: [
          {
            type: "text" as const,
            text: JSON.stringify({
              status: "approved",
              candidate_id,
              key: candidate.key,
            }),
          },
        ],
      };
    },
  );
}
